homeassistant/components/water_heater/wink.py

```diff
@@ -0,0 +1,136 @@
+"""
+Support for Wink water heaters.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/water_heater.wink/
+"""
+import logging
+
+from homeassistant.components.water_heater import (
+    ATTR_TEMPERATURE, STATE_ECO, STATE_ELECTRIC,
+    STATE_PERFORMANCE, SUPPORT_AWAY_MODE, STATE_HEAT_PUMP,
+    STATE_GAS, STATE_HIGH_DEMAND,
+    SUPPORT_OPERATION_MODE, SUPPORT_TARGET_TEMPERATURE,
+    WaterHeaterDevice)
+from homeassistant.components.wink import DOMAIN, WinkDevice
+from homeassistant.const import (
+    STATE_OFF, STATE_UNKNOWN, TEMP_CELSIUS)
+
+_LOGGER = logging.getLogger(__name__)
+
+SUPPORT_FLAGS_HEATER = (SUPPORT_TARGET_TEMPERATURE | SUPPORT_OPERATION_MODE |
+                        SUPPORT_AWAY_MODE)
+
+ATTR_RHEEM_TYPE = 'rheem_type'
+ATTR_VACATION_MODE = 'vacation_mode'
+
+HA_STATE_TO_WINK = {
+    STATE_ECO: 'eco',
+    STATE_ELECTRIC: 'electric_only',
+    STATE_GAS: 'gas',
+    STATE_HEAT_PUMP: 'heat_pump',
+    STATE_HIGH_DEMAND: 'high_demand',
+    STATE_OFF: 'off',
+    STATE_PERFORMANCE: 'performance',
+}
+
+WINK_STATE_TO_HA = {value: key for key, value in HA_STATE_TO_WINK.items()}
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the Wink water heater devices."""
+    import pywink
+    for water_heater in pywink.get_water_heaters():
+        _id = water_heater.object_id() + water_heater.name()
+        if _id not in hass.data[DOMAIN]['unique_ids']:
+            add_entities([WinkWaterHeater(water_heater, hass)])
```

Comment:
Passing in `hass` to the entity is usually a sign that we're doing something else wrong. Here it looks like we're subscribing to updates before the entity has been added to home assistant in `WinkDevice`.

Reply:
Hmm not too sure I know what you mean. How would I not pass in `hass`?  I think you are correct, we are subscribing to updates before it gets added, but is that an issue? This is how all of the Wink components work.

Reply:
If an update would come in before the entity has been added to home assistant there would be an error since the entity isn't set up properly yet. This is why subscribing for updates or registering callbacks etc should be done in `async_added_to_hass`.

`hass` will be set on the entity when it has been added to home assistant:
https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#5-entity